pkg/gatewayserver/io/rtts.go

```diff
@@ -20,55 +20,69 @@ import (
 	"time"
 )
 
+type rttItem struct {
+	d time.Duration
+	t time.Time
+}
+
 type rtts struct {
 	count int
+	ttl   time.Duration
 	mu    sync.RWMutex
-	items []time.Duration
+	items []rttItem
 }
 
-func newRTTs(count int) *rtts {
+func newRTTs(count int, ttl time.Duration) *rtts {
 	return &rtts{
 		count: count,
-		items: make([]time.Duration, 0, count+1),
+		ttl:   ttl,
+		items: make([]rttItem, 0, count+1),
 	}
 }
 
 // Record records the given round-trip time.
-func (r *rtts) Record(d time.Duration) {
+func (r *rtts) Record(d time.Duration, t time.Time) {
 	r.mu.Lock()
-	r.items = append(r.items, d)
+	r.items = append(r.items, rttItem{d, t})
 	if len(r.items) > r.count {
-		r.items = append(r.items[:0], r.items[len(r.items)-r.count:]...)
+		r.items = append(r.items[:0], r.items[1:]...)
 	}
 	r.mu.Unlock()
 }
 
 // Last returns the last measured round-trip time.
-func (r *rtts) Last() (time.Duration, bool) {
+func (r *rtts) Last(ref time.Time) (time.Duration, bool) {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
-	if len(r.items) == 0 {
+	if len(r.items) == 0 || ref.Sub(r.items[len(r.items)-1].t) > r.ttl {
```

Comment:
Wouldn't this cause a seg fault if `r.items` is empty?
The logical OR `||` has the least precedence here with both brackets and the relational `>` evaluated first before the OR condition.

Reply:
Evaluation is left-to-right here

Reply:
Ok that's actually counterintuitive. If expressions get evaluated L->R regardless, then operator precedence has no effect.

Reply:
It does, operator precedence makes us avoid having to write `a || (b > c)` because `>` is higher than `||`, `b > c` is a group, but the `a` is evaluated before `b > c`

Reply:
Yes but without using brackets `(b > c)`  there's no way to determine which is the group. Both `(a || b) > c` and `a || (b > c)` are equally likely and this is where precedence comes in. But anyway this is off topic.

Reply:
https://kuree.gitbooks.io/the-go-programming-language-report/content/22/text.html